server/block/campfire.go

```diff
@@ -0,0 +1,267 @@
+	return false
+}
+
+// Splash checks to see if the fire was splashed by a bottle and then extinguishes itself
+func (c Campfire) Splash(pos cube.Pos, p *entity.SplashPotion) {
+	if p.Type() != potion.Water() {
+		return
+	}
+	w := p.World()
+	c.Extinguished = true
+	p.World().PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+	w.SetBlock(pos, c, nil)
+}
+
+// BreakInfo ...
+func (c Campfire) BreakInfo() BreakInfo {
+	return newBreakInfo(2, alwaysHarvestable, axeEffective, func(t item.Tool, enchantments []item.Enchantment) []item.Stack {
+		var drops []item.Stack
+		if hasSilkTouch(enchantments) {
+			drops = append(drops, item.NewStack(c, 1))
+		} else {
+			switch c.Type {
+			case NormalFire():
+				drops = append(drops, item.NewStack(item.Charcoal{}, 2))
+			case SoulFire():
+				drops = append(drops, item.NewStack(SoulSoil{}, 1))
+			default:
+				panic("invalid fire type")
+			}
+		}
+		for _, v := range c.Items {
+			if !v.Item.Empty() {
+				drops = append(drops, v.Item)
+			}
+		}
+		return drops
+	})
+}
+
+// LightEmissionLevel ...
+func (c Campfire) LightEmissionLevel() uint8 {
+	if c.Extinguished {
+		return 0
+	}
+	switch c.Type {
+	case NormalFire():
+		return 15
+	case SoulFire():
+		return 10
+	}
+	panic("invalid fire type")
+}
+
+// Activate ...
+func (c Campfire) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.User, ctx *item.UseContext) bool {
+	if held, _ := u.HeldItems(); !held.Empty() {
+		if _, ok := held.Item().(item.FlintAndSteel); ok {
+			w.PlaySound(pos.Vec3(), sound.Ignite{})
+			c.Extinguished = false
+			w.SetBlock(pos, c, nil)
+			return true
+		}
+		if rawFood, ok := held.Item().(item.Smeltable); ok && rawFood.SmeltInfo().Food {
+			for i, it := range c.Items {
+				if it.Item.Empty() {
+					c.Items[i] = CampfireItem{
+						Item: held.Grow(-held.Count() + 1),
+						Time: 600,
+					}
+					ctx.SubtractFromCount(1)
+					w.PlaySound(pos.Vec3Centre(), sound.AddItem{})
+					w.SetBlock(pos, c, nil)
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// UseOnBlock ...
+func (c Campfire) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	if _, ok := w.Block(pos).(Campfire); ok && face == cube.FaceUp {
+		return false
+	}
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return
+	}
+	c.Facing = user.Facing().Opposite()
+	place(w, pos, c, user, ctx)
+	return placed(ctx)
+}
+
+// Tick is called to cook the items within the campfire
+func (c Campfire) Tick(_ int64, pos cube.Pos, w *world.World) {
+	if !c.Extinguished && rand.Float64() <= 0.016 { // Every three or so seconds.
+		w.PlaySound(pos.Vec3Centre(), sound.CampfireCrackle{})
+	}
+	if !c.Extinguished {
+		// if the campfire is water logged we extinguish it
+		if _, ok := w.Liquid(pos); ok {
+			c.Extinguished = true
+			w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+			w.SetBlock(pos, c, nil)
```

Comment:
Should `return` here as cooking progress should not increase after being extinguished

Reply:
Although this should probably be handled on neighbour updates? Check the sea pickle block

Reply:
Would neighbors updating also work for blocks on a different world layer?

Reply:
Yes, that's how the sea pickle blocks work

Reply:
done!